scipy/sparse/linalg/_dsolve/linsolve.py

```diff
@@ -651,93 +649,94 @@ def spsolve_triangular(A, b, lower=True, overwrite_A=False, overwrite_b=False,
     Examples
     --------
     >>> import numpy as np
-    >>> from scipy.sparse import csr_matrix
+    >>> from scipy.sparse import csc_array
     >>> from scipy.sparse.linalg import spsolve_triangular
-    >>> A = csr_matrix([[3, 0, 0], [1, -1, 0], [2, 0, 1]], dtype=float)
+    >>> A = csc_array([[3, 0, 0], [1, -1, 0], [2, 0, 1]], dtype=float)
     >>> B = np.array([[2, 0], [-1, 0], [2, 0]], dtype=float)
     >>> x = spsolve_triangular(A, B)
     >>> np.allclose(A.dot(x), B)
     True
     """
 
     if is_pydata_spmatrix(A):
-        A = A.to_scipy_sparse().tocsr()
-
-    # Check the input for correct type and format.
-    if not (issparse(A) and A.format == "csr"):
-        warn('CSR matrix format is required. Converting to CSR matrix.',
+        A = A.to_scipy_sparse().tocsc()
+    
+    trans = "N"
+    if issparse(A) and A.format == "csr":
+        A = A.T
+        trans = "T"
+        lower = not lower
+
+    if not (issparse(A) and A.format == "csc"): 
+        warn('CSC or CSR matrix format is required. Converting to CSC matrix.',
              SparseEfficiencyWarning)
-        A = csr_matrix(A)
+        A = csc_matrix(A)
     elif not overwrite_A:
         A = A.copy()
 
-    if A.shape[0] != A.shape[1]:
+
+    M, N = A.shape
+    if M != N:
         raise ValueError(
             f'A must be a square matrix but its shape is {A.shape}.')
 
+    if unit_diagonal:
+        A = A.tolil()
+        A.setdiag(1)
+        A = A.tocsc()
```

Comment:
That's the CSC->LIL->CSC conversion mentioned in https://github.com/scipy/scipy/pull/17924#discussion_r1390250375.

Reply:
This looks fine.

Reply:
My 2 cents: some quick benchmarking reveals that the CSC->LIL->CSC dance is actually faster (for the shapes / densities I tested) than calling CSC's `setdiag(1)` directly.

That said, I view this to be a performance bug for CSC format, and ideally we can write a quick PR to at least make the direct `setdiag()` call no less efficient than the format conversion approach.

Reply:
I can partially confirm this observation on my x86_64 Ubuntu system with Python 3.10.12 and SciPy 1.11.0 using 
[benchmark.py.zip](https://github.com/scipy/scipy/files/14008153/benchmark.py.zip). Converting to LIL, setting the diagonal there and then converting back to CSC is faster than calling `setdiag` on the CSC matrix directly in setups with low size or low density:

Density 0.001:

![plot_0_001](https://github.com/scipy/scipy/assets/72806890/797ff31e-ef6b-4781-8d13-6e38cdfc08c1)

Density 0.01:

![plot_0_01](https://github.com/scipy/scipy/assets/72806890/830f2f3f-3396-4a28-ae1d-8fb04ea654c2)

Density 0.1:

![plot_0_1](https://github.com/scipy/scipy/assets/72806890/6a64afd9-5146-4912-aa44-1f1294a00d28)

@perimosocordiae Thanks for reporting this, I think that it would be reasonable in this PR replace the present CSC->LIL->CSC conversion by a CSC `setdiag` call (which is sometimes slower but not massively) and see how to suppress the warning message. You could then open an issue on the performance bug but I'd prefer to avoid making this part of this PR. What do you think?

Reply:
Thanks for the analysis! This aligns with what I observed locally as well.

+1 to using CSC.setdiag() with the warning suppressed, and I'll take care of opening the issue to track the performance bug.

Reply:
I just opened gh-19943 to track it.

Reply:
Thanks! b706a7d3c0f86c6771c3f532bb667bd2bf080b50 replaces the CSC->LIL->CSC conversion with CSC.setdiag with the SparseEfficiencyWarning suppressed.